lib/spack/spack/directives.py

```diff
@@ -545,7 +551,29 @@ def depends_on(spec, when=None, type=default_deptype, patches=None):
     """
 
     def _execute_depends_on(pkg):
-        _depends_on(pkg, spec, when=when, type=type, patches=patches)
+        executed = False
+        if version_translator:
+            # Produce a different dependency for every package version
+            # for which version_translator returns an answer that alters
+            # the spec.
+            for pkg_version in pkg.versions:
+                version_ish = version_translator(pkg_version)
+                if version_ish is not None:
+                    constrained_spec = spack.spec.Spec(spec)
+                    if constrained_spec.constrain(f"@{version_ish}"):
```

Comment:
I don't understand what you mean, here: "This method always returns ..."

The `depends_on()` directive here always returns the `_execute_depends_on` function. `_execute_depends_on()` always results in at least one execution of the `_depends_on()` function, just as did the pre-PR implementation.

`version_translator()`—if defined by the user—may return anything, but the intention is for it to return either None or something that converts to a string which is a valid version part of a version constraint (hence `version_ish`). This could be a valid `Version` instance (which could be a standard version, range, list, or any other valid version type), or it could be a string.